src/main/java/cws/k8s/scheduler/memory/TaskScaler.java

```diff
@@ -0,0 +1,346 @@
+        if (task.wasSuccessfullyExecuted()) {
+            peakRss = NfTrace.getNfPeakRss(task);
+            peakVmem = NfTrace.getNfPeakVmem(task);
+            realtime = NfTrace.getNfRealTime(task);
+        } else {
+            peakRss = BigDecimal.ZERO;
+            peakVmem = BigDecimal.ZERO;
+            realtime = 0;
+            // when a task has failed, we put it on the blacklist, so we will not tamper it again
+            this.blacklist.add(task.getConfig().getName());
+        }
+        // @formatter:off
+        Observation o = Observation.builder()
+                .task( task.getConfig().getTask() )
+                .taskName( task.getConfig().getName() )
+                .success( task.wasSuccessfullyExecuted() )
+                .inputSize( task.getInputSize() )
+                .ramRequest( task.getPod().getRequest().getRam() )
+                .peakVmem(peakVmem)
+                .peakRss(peakRss)
+                .realtime(realtime)
+                .node(task.getNode().getName())
+                .build();
+        // @formatter:on
+        log.info("taskWasFinished, observation={}", o);
+        memoryPredictor.addObservation(o);
+        statistics.addObservation(o);
+
+        // Note: this is a workaround, because the SchedulerConfig does not contain the baseDir
+        if (statistics.baseDir == null) {
+            statistics.baseDir = task.getWorkingDir().substring(0, task.getWorkingDir().lastIndexOf("work"));
+        }
+    }
+
+    public synchronized void beforeTasksScheduled(final List<Task> unscheduledTasks) {
+        if (!active) {
+            return;
+        }
+        log.debug("--- unscheduledTasks BEGIN ---");
+        for (Task t : unscheduledTasks) {
+            log.debug("1 unscheduledTask: {} {} {}", t.getConfig().getTask(), t.getConfig().getName(),
+                    t.getPod().getRequest());
+
+            // if task is already blacklisted, don't touch it again
+            if (this.blacklist.contains(t.getConfig().getName())) {
+                continue;
+            }
+            
+            // if task had no memory request set, it cannot be changed
+            BigDecimal taskRequest = t.getPod().getRequest().getRam();
+            if (taskRequest.compareTo(BigDecimal.ZERO) == 0) {
+                log.info("cannot change task {}, because it had no prior requirements", t.toString());
+                continue;
+            }
+            
+            BigDecimal newRequestValue = null;
+                        
+            // sanity check for Nextflow provided value
+            if (taskRequest.compareTo(this.maxRequest) > 0) {
+                // this would never get scheduled and CWS will get stuck, so we take the liberty to lower the value
+                newRequestValue = this.maxRequest.subtract(BigDecimal.valueOf(1l*1024*1024));
+                log.warn("nextflow request exceeds maximal cluster allocatable capacity, request was reduced by TaskScaler");
+            }
```

Comment:
As I wrote before, the maxRequest depends on the current state of determining. Typically you set this to the maximum memory on any node.  Not the maximum available. I think we should ignore this here completely, as this is the user defined value and thus the user's fault.